hikari/utilities/event_stream.py

```diff
@@ -0,0 +1,244 @@
+
+        await method(event)
+
+    return call_weak_method
+
+
+class EventStream(Streamer[EventT]):
+    """An implementation of an event `Streamer` class.
+
+    !!! note
+        While calling `EventStream.filter` on an active "opened" event stream
+        will return a wrapping lazy iterator, calling it on an inactive "closed"
+        event stream will return the event stream and add the given predicates
+        to the streamer.
+    """
+
+    __slots__ = ("_active", "_app", "_event_type", "_filters", "_queue", "_registered_listener", "_timeout")
+
+    def __init__(
+        self,
+        app: traits.BotAware,
+        event_type: typing.Type[EventT],
+        *,
+        timeout: typing.Union[float, int, None],
+        limit: typing.Optional[int] = None,
+    ) -> None:
+        self._app = app
+        self._active = False
+        self._event_type = event_type
+        self._filters: iterators.All[EventT] = iterators.All(())
+        # We accept `None` to represent unlimited here to be consistent with how `None` is already used to represent
+        # unlimited for timeout in other places.
+        self._queue: asyncio.Queue[EventT] = asyncio.Queue(limit or 0)
+        self._registered_listener: typing.Optional[
+            typing.Callable[[EventT], typing.Coroutine[typing.Any, typing.Any, None]]
+        ] = None
+        # The registered wrapping function for the weak ref to this class's _listener method.
+        self._timeout = timeout
+
+    async def _listener(self, event: EventT) -> None:
+        if not self._filters(event):
+            return
+
+        try:
+            self._queue.put_nowait(event)
+        except asyncio.QueueFull:
+            pass
+
+    async def __anext__(self) -> EventT:
+        if not self._active:
+            raise TypeError("stream must be started with `await with` before entering it")
+
+        try:
+            return await asyncio.wait_for(self._queue.get(), timeout=self._timeout)
+        except asyncio.TimeoutError:
+            raise StopAsyncIteration from None
+
+    async def _await_all(self) -> typing.Sequence[EventT]:
+        await self.open()
+        result = [event async for event in self]
+        await self.close()
+        return result
+
+    def __await__(self) -> typing.Generator[None, None, typing.Sequence[EventT]]:
+        return self._await_all().__await__()
+
+    def __del__(self) -> None:
+        # For the sake of protecting highly intelligent people who forget to close this, we register the event listener
+        # with a weakref then try to close this on deletion. While this may lead to their consoles being spammed, this
+        # is a small price to pay as it'll be way more obvious what's wrong than if we just left them with a vague
+        # ominous memory leak.
+        if self._active:
+            _LOGGER.warning("active %r streamer fell out of scope before being closed", self._event_type.__name__)
+            asyncio.ensure_future(self.close())
```

Comment:
We should check that the event loop is running first.

If it isn't running, this will spit out random errors, and you don't really care if the loop isn't running that you don't close this properly as the consumer will be dead anyway.

Reply:
What errors will it spit out if the loop isn't running?

Reply:
```py
>>> import asyncio
>>> 
>>> async def foo():
...   pass
... 
>>> asyncio.run(foo())
>>> 
>>> asyncio.ensure_future(foo())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.8/asyncio/tasks.py", line 660, in ensure_future
    loop = events.get_event_loop()
  File "/usr/lib/python3.8/asyncio/events.py", line 639, in get_event_loop
    raise RuntimeError('There is no current event loop in thread %r.'
RuntimeError: There is no current event loop in thread 'MainThread'.
```

Reply:
there's no good way to check in that situation, you'd get the exact same behaviour from asyncio.get_event_loop so best just to catch there

Reply:
my point is, that RuntimeError should be caught.

```suggestion
            try:
                asyncio.ensure_future(self.close())
            except RuntimeError:
                # event loop has closed, nothing can be streamed now.
                ...
```